<a href="https://colab.research.google.com/github/data602sps/assignments/blob/master/05_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7:**


# Introduction:
The dataset I am using for this assignment is a IMDb movies dataset containing infromation about movies' name, genre, release year, rating, runtime, gross earning, director and actor names. I have collected this dataset from kaggle.com [https://www.kaggle.com/datasets/bharatnatrayn/movies-dataset-for-feature-extracion-prediction], and then stored on my GitHub [https://github.com/Raji030/data602_project_data/blob/main/movies.csv]. 

I have chosen this dataset because I am a great consumer of movies. It is a great source of entertainment to me and it also helps me to get relieved from stresses of everyday life. Morever, I always find it interesting to understand the consumers' culture and the factors that drive a movie towards success.I am also curious about the difference between the viewer popularity and total gross.

______________
# Data exploration part:
In this part, I will import the dataset into my .ipynb, create dataframes, and explore the data including the following: 

* Summary statistics means, medians, quartiles, 
* Missing value information
* Any other relevant information about the dataset.  



# Import dataset using pandas 

In [271]:
import pandas as pd
url="https://raw.githubusercontent.com/Raji030/data602_project_data/main/movies.csv"
df=pd.read_csv(url)
df.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN

# Summary statistics:

In [272]:
df.describe()

RATING      RunTime
count  8179.000000  7041.000000
mean      6.921176    68.688539
std       1.220232    47.258056
min       1.100000     1.000000
25%       6.200000    36.000000
50%       7.100000    60.000000
75%       7.800000    95.000000
max       9.900000   853.000000

# Missing value information:

In [273]:
missing_values = df.isnull().sum()
print(missing_values)

MOVIES         0
YEAR         644
GENRE         80
RATING      1820
ONE-LINE       0
STARS          0
VOTES       1820
RunTime     2958
Gross       9539
dtype: int64


# Other relevant information:
Other relevant information include rows and columns number, data type for each column, and the columns' names.

In [274]:
# See data shape i.e. row and column numbers
df.shape

(9999, 9)

In [275]:
# See data type in different columns
df.dtypes

MOVIES       object
YEAR         object
GENRE        object
RATING      float64
ONE-LINE     object
STARS        object
VOTES        object
RunTime     float64
Gross        object
dtype: object

In [276]:
# See column names
df.columns.tolist()

['MOVIES',
 'YEAR',
 'GENRE',
 'RATING',
 'ONE-LINE',
 'STARS',
 'VOTES',
 'RunTime',
 'Gross']

# Data wrangling part:
At first, I will create a subset of the original data and then perform the following: 

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort the data based on multiple variables. 

8. Filter the data based on some condition. 

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group the dataset by one column, and get the mean, min, and max values by group. 
  * Groupby()
  * agg() or .apply()

12. Group the dataset by two columns and then sort the aggregated results within the groups. 

# Create subset:

In [277]:
# Subset,df1 of dataframe by dropping 3rd, 5th and 6th columns
df1 = df.drop(df.columns[[2,4,5]], axis=1)
df1.head()

MOVIES         YEAR  RATING    VOTES  RunTime  \
0                        Blood Red Sky       (2021)     6.1   21,062    121.0   
1  Masters of the Universe: Revelation     (2021– )     5.0   17,870     25.0   
2                     The Walking Dead  (2010–2022)     8.2  885,805     44.0   
3                       Rick and Morty     (2013– )     9.2  414,849     23.0   
4                      Army of Thieves       (2021)     NaN      NaN      NaN   

  Gross  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

# Modify multiple column names:

In [278]:
df1 = df1.rename(columns={'MOVIES':'movies', 'YEAR':'year', 'RATING':'rating', 'VOTES':'votes', 'RunTime':'runtime', 'Gross':'gross'})
df1.head()

movies         year  rating    votes  runtime  \
0                        Blood Red Sky       (2021)     6.1   21,062    121.0   
1  Masters of the Universe: Revelation     (2021– )     5.0   17,870     25.0   
2                     The Walking Dead  (2010–2022)     8.2  885,805     44.0   
3                       Rick and Morty     (2013– )     9.2  414,849     23.0   
4                      Army of Thieves       (2021)     NaN      NaN      NaN   

  gross  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN


# Convert different variables' data type into correct structure i.e. from non-numeric to numeric data type:

In [279]:
# Extract only single year in year column
df1['year'] = df1['year'].str.extract(r'(\d{4})')

# Convert year column to numeric data type
df1['year'] = pd.to_numeric(df1['year'])

# Remove commas from votes column
df1['votes'] = df1['votes'].replace(',', '', regex=True)

# Convert votes column to numeric data type
df1['votes'] = pd.to_numeric(df1['votes'])

# Extract numeric value in gross column
df1['gross'] = df1['gross'].str.extract('(\d+\.\d+|\d+)')

# Convert gross column to numeric
df1['gross'] = pd.to_numeric(df1['gross'])

# Check data types 
print(df1.dtypes)

df1.head()


movies      object
year       float64
rating     float64
votes      float64
runtime    float64
gross      float64
dtype: object


movies    year  rating     votes  runtime  \
0                        Blood Red Sky  2021.0     6.1   21062.0    121.0   
1  Masters of the Universe: Revelation  2021.0     5.0   17870.0     25.0   
2                     The Walking Dead  2010.0     8.2  885805.0     44.0   
3                       Rick and Morty  2013.0     9.2  414849.0     23.0   
4                      Army of Thieves  2021.0     NaN       NaN      NaN   

   gross  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

# Fix missing and invalid values in data:

In [280]:
# Fill missing values with mean in rating, votes and runtime columns
df1['rating'].fillna(df1['rating'].mean(), inplace=True)
df1['votes'].fillna(df1['votes'].mean(), inplace=True)
df1['runtime'].fillna(df1['runtime'].mean(), inplace=True)

# Replace invalid values i.e. remove dollar signs and commas from the 'gross' column
df1['gross'] = df1['gross'].replace({'\$': '', ',': ''}, regex=True)
#df1['gross'].fillna(df1['gross'].mean(), inplace=True)

# Drop missing values in year and gross columns
df1.dropna(subset=['year', 'gross'], inplace=True)

# Drop duplicates
df1.drop_duplicates(inplace=True)
df1.head()

# Check if any missing value is still present
missing_values1 = df1.isnull().sum()
print(missing_values1)

movies     0
year       0
rating     0
votes      0
runtime    0
gross      0
dtype: int64


# Create new columns based on existing columns or calculations:

In [281]:
# Create a new column 'year_release' by converting year column to integer type 
df1['year_release'] = df1['year'].astype(int)

# Create new column named gross_earning 
df1['gross_earning'] = df1['gross']
df1.head()

movies    year  rating  \
77                              The Hitman's Bodyguard  2017.0     6.9   
85                                       Jurassic Park  1993.0     8.1   
95                                       Don't Breathe  2016.0     7.1   
111  The Lord of the Rings: The Fellowship of the Ring  2001.0     8.8   
125                                        Escape Room  2019.0     6.4   

         votes  runtime   gross  year_release  gross_earning  
77    205979.0    118.0   75.47          2017          75.47  
85    897444.0    127.0  402.45          1993         402.45  
95    237601.0     88.0   89.22          2016          89.22  
111  1713028.0    178.0  315.54          2001         315.54  
125    99351.0     99.0   57.01          2019          57.01

# Drop column(s) from dataset:

In [282]:
# Drop year and gross columns
df1 = df1.drop(["year","gross"], axis=1)
df1.head()

movies  rating      votes  \
77                              The Hitman's Bodyguard     6.9   205979.0   
85                                       Jurassic Park     8.1   897444.0   
95                                       Don't Breathe     7.1   237601.0   
111  The Lord of the Rings: The Fellowship of the Ring     8.8  1713028.0   
125                                        Escape Room     6.4    99351.0   

     runtime  year_release  gross_earning  
77     118.0          2017          75.47  
85     127.0          1993         402.45  
95      88.0          2016          89.22  
111    178.0          2001         315.54  
125     99.0          2019          57.01

# Drop a row(s) from dataset:

In [283]:
# Check if any column has a zero value
if (df1 == 0).any(axis=0).any():
    print("At least one column has a zero value")
else:
    print("No column has a zero value")

# Drop rows containing at least one zero value in any column
df1 = df1[~(df1==0).any(axis=1)]

# Check again if any column has a zero value
if (df1 == 0).any(axis=0).any():
    print("At least one column has a zero value")
else:
    print("No column has a zero value")

df1.head()

At least one column has a zero value
No column has a zero value


movies  rating      votes  \
77                              The Hitman's Bodyguard     6.9   205979.0   
85                                       Jurassic Park     8.1   897444.0   
95                                       Don't Breathe     7.1   237601.0   
111  The Lord of the Rings: The Fellowship of the Ring     8.8  1713028.0   
125                                        Escape Room     6.4    99351.0   

     runtime  year_release  gross_earning  
77     118.0          2017          75.47  
85     127.0          1993         402.45  
95      88.0          2016          89.22  
111    178.0          2001         315.54  
125     99.0          2019          57.01

# Sort the data based on multiple variables:

In [284]:
# Sort data based on gross eraning and rating variables
df1 = df1.sort_values(['gross_earning','rating'], ascending=[True,False])
df1.head()

movies  rating    votes  runtime  year_release  \
1729                    Blue Jay     7.3  15777.0     80.0          2016   
3204              Aala Kaf Ifrit     7.0   1912.0    100.0          2017   
4196         The Sunshine Makers     7.0   1344.0    101.0          2015   
5290  The Search for General Tso     6.9   1726.0     71.0          2014   
6056              Theo Who Lived     6.8    111.0     86.0          2016   

      gross_earning  
1729           0.01  
3204           0.01  
4196           0.01  
5290           0.01  
6056           0.01

# Filter the data based on some condition:

In [285]:
# Filter the movies that have a rating greater than 7 and gross earning greater than 400 million dollar
df2 = df1[(df1["rating"] > 7) & (df1["gross_earning"] > 400)]
df2.head()

movies  rating     votes  runtime  year_release  \
351                      Frozen     7.4  586769.0    102.0          2013   
85                Jurassic Park     8.1  897444.0    127.0          1993   
196                  Spider-Man     7.3  709684.0    121.0          2002   
226  Captain America: Civil War     7.8  697750.0    147.0          2016   
578                Finding Dory     7.3  250953.0     97.0          2016   

     gross_earning  
351         400.74  
85          402.45  
196         403.71  
226         408.08  
578         486.30

# Convert all the string values to upper or lower cases in one column:

In [286]:
# Convert 'movies' column to lowercase:
df1['movies'] = df1['movies'].str.lower()
df1.head()

movies  rating    votes  runtime  year_release  \
1729                    blue jay     7.3  15777.0     80.0          2016   
3204              aala kaf ifrit     7.0   1912.0    100.0          2017   
4196         the sunshine makers     7.0   1344.0    101.0          2015   
5290  the search for general tso     6.9   1726.0     71.0          2014   
6056              theo who lived     6.8    111.0     86.0          2016   

      gross_earning  
1729           0.01  
3204           0.01  
4196           0.01  
5290           0.01  
6056           0.01

# Check whether numeric values are present in a given column in the dataframe:

In [287]:
# Check if the gross_earning column has all numeric values
if pd.api.types.is_numeric_dtype(df1["gross_earning"]):
    print("All values in gross_earning column are numeric")
else:
    print("The column contains non-numeric values")

All values in gross_earning column are numeric


# Group the dataset by one column, and get the mean, min, and max values by group:

In [288]:
# Group the dataset by movies column and get the mean, min, and max values by group
df3=df1.groupby("movies").agg({"rating": ["mean", "min", "max"]})
df3.head()

rating          
                        mean  min  max
movies                                
101 dalmatians           5.7  5.7  5.7
127 hours                7.5  7.5  7.5
17 filles                6.0  6.0  6.0
2001: a space odyssey    8.3  8.3  8.3
300                      7.6  7.6  7.6

# Group the dataset by two columns and then sort the aggregated results within the groups:

In [289]:
# Group the dataset by movies and year columns and get the mean, min, and max values by group
df4 = df1.groupby(["movies", "year_release"]).agg({"rating": ["mean", "min", "max"], "gross_earning": ["mean", "min", "max"]})

# Sort the aggregated results within the groups
df5 = df4.sort_values(by=[("rating", "mean"), ("gross_earning", "mean")], ascending=[False, False])

df5.head()

rating       \
                                                                 mean  min   
movies                                            year_release               
the lord of the rings: the return of the king     2003            8.9  8.9   
the lord of the rings: the fellowship of the ring 2001            8.8  8.8   
the lord of the rings: the two towers             2002            8.7  8.7   
sen to chihiro no kamikakushi                     2001            8.6  8.6   
be here now                                       2015            8.6  8.6   

                                                                     \
                                                                max   
movies                                            year_release        
the lord of the rings: the return of the king     2003          8.9   
the lord of the rings: the fellowship of the ring 2001          8.8   
the lord of the rings: the two towers             2002          8.7   
sen to chihiro no kamikakushi                     2001          8.6   
be here now                                       2015          8.6   

                                                               gross_earning  \
                                                                        mean   
movies                                            year_release                 
the lord of the rings: the return of the king     2003                377.85   
the lord of the rings: the fellowship of the ring 2001                315.54   
the lord of the rings: the two towers             2002                342.55   
sen to chihiro no kamikakushi                     2001                 10.06   
be here now                                       2015                  0.02   

                                                                                
                                                                   min     max  
movies                                            year_release                  
the lord of the rings: the return of the king     2003          377.85  377.85  
the lord of the rings: the fellowship of the ring 2001          315.54  315.54  
the lord of the rings: the two towers             2002          342.55  342.55  
sen to chihiro no kamikakushi                     2001           10.06   10.06  
be here now                                       2015            0.02    0.02

# Conclusion: 
After exploring the dataset, I have found that the dataset has 9,999 rows and 9 columns. The columns include information such as the movie title, year of release, genre, duration, rating, votes, director, actors, and gross income. The dataset has missing values in the year, genre, rating,votes, run time and gross columns, and the data type for year, votes, and gross earning cloumns are not in appropriate data type. The mean of the movie rating is 6.92 and the mean of the movie run time is 68.69 minutes. In this dataset, the target variable is not clearly defined. I could potentially use the gross income as a proxy for movie's success and explore its distribution along with other movie features' exploratory data analysis and visualization by using different plots. Additionally, I will further explore the dataset to find out the following questions: 
(1) Which is the most successful movie based on gross earning in the dataset? (2) Is the movie that grosses the most also the most popular among viewers? 
(3) Which year had the longest run time of movies? 
(4) What are the top 10 movies based on rating frequency?